packages/gatsby/src/datastore/lmdb/query/common.ts

```diff
@@ -38,35 +38,35 @@ export function resolveFieldValue(
   return getValueAt(node, dottedFieldPath)
 }
 
-export function shouldFilter(
+export function matchesFilter(
   filter: IDbFilterStatement,
   fieldValue: unknown
 ): boolean {
-  const value = filter.value
-
   switch (filter.comparator) {
     case DbComparator.EQ:
-      return value === null ? value == fieldValue : value === fieldValue
+      return filter.value === null
+        ? filter.value == fieldValue
+        : filter.value === fieldValue
     case DbComparator.IN: {
-      const arr = Array.isArray(value) ? value : [value]
+      const arr = Array.isArray(filter.value) ? filter.value : [filter.value]
       return arr.some(v => (v === null ? v == fieldValue : v === fieldValue))
     }
     case DbComparator.GT:
-      return compareKey(value, fieldValue) > 0
+      return compareKey(fieldValue, filter.value) > 0
```

Comment:
This line is essentially the fix - just swapping arguments as the comparison was incorrect here. 
Other changes are just added tests and minor rename/refactors.